src/drivers/mediatek/mt8186/timer.c

```diff
@@ -0,0 +1,213 @@
+// SPDX-License-Identifier: BSD-3-Clause
+//
+// Copyright(c) 2022 Mediatek. All rights reserved.
+//
+// Author: Allen-KH Cheng <allen-kh.cheng@mediatek.com>
+//         Tinghan Shen <tinghan.shen@mediatek.com>
+
+#include <sof/audio/component_ext.h>
+#include <sof/drivers/interrupt.h>
+#include <sof/lib/memory.h>
+#include <sof/platform.h>
+#include <platform/drivers/timer.h>
+#include <platform/drivers/interrupt.h>
+#include <ipc/stream.h>
+#include <errno.h>
+#include <stdint.h>
+
+void platform_timer_start(struct timer *timer)
+{
+	/* set 26m clksrc */
+	io_reg_update_bits(MTK_TIMER_CON(timer->id), MTK_TIMER_ENABLE_BIT, 0);
+	io_reg_update_bits(MTK_TIMER_CON(timer->id), MTK_TIMER_CLKSRC_BIT,
+			   MTK_TIMER_CLK_SRC_CLK_26M);
+
+	/* enable timer */
+	io_reg_update_bits(MTK_TIMER_CON(timer->id), MTK_TIMER_ENABLE_BIT,
+			   MTK_TIMER_ENABLE_BIT);
+
+	/* enable system boot time timer */
+	io_reg_update_bits(MTK_OSTIMER_CON, MTK_OSTIMER_EN_BIT,
+			   MTK_OSTIMER_EN_BIT);
+}
+
+void platform_timer_stop(struct timer *timer)
+{
+	if (timer->id > NR_TMRS)
+		return;
+
+	io_reg_update_bits(MTK_TIMER_CON(timer->id), MTK_TIMER_ENABLE_BIT, 0);
+	io_reg_update_bits(MTK_TIMER_IRQ_ACK(timer->id), MTK_TIMER_IRQ_ENABLE, 0);
+}
+
+/* IRQs off in arch_timer_get_system() */
+uint64_t platform_timer_get_atomic(struct timer *timer)
+{
+	return platform_timer_get(timer);
+}
+
+int64_t platform_timer_set(struct timer *timer, uint64_t ticks)
+{
+	uint64_t time;
+	uint32_t flags;
+	uint32_t low, high;
+	uint32_t ticks_set;
+
+	if (timer->id > NR_TMRS)
+		return -EINVAL;
+
+	flags = arch_interrupt_global_disable();
+
+	low = io_reg_read(MTK_OSTIMER_CUR_L);
+	high = io_reg_read(MTK_OSTIMER_CUR_H);
+
+	/* ostimer 13M counter to 26M interrupt */
+	time = (((uint64_t)high << 32) | low) << 1;
+
+	ticks_set = (ticks > time) ? ticks - time : UINT64_MAX - time +  ticks;
+
+	timer->hitimeout = ticks >> 32;
+	timer->lowtimeout = ticks_set;
+
+	/* disable timer before config it */
+	io_reg_update_bits(MTK_TIMER_CON(timer->id), MTK_TIMER_ENABLE_BIT, 0);
+
+	/* timer reset value */
+	io_reg_write(MTK_TIMER_RST_VAL(timer->id), ticks_set);
```

Comment:
this is in units, half of what was read in line 61 from `MTK_OSTIMER_CUR_L` because of multiplication in line 65, is this correct?

Reply:
This is intended. The underlying HW is a 32-bit count-down timer.
It's used to set the difference between the current tick and next tick triggering reschedule.